packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -155,6 +155,15 @@ export default createRule<Options, MessageIds>({
             continue;
           }
 
+          // `undefined` should cover the "missing" undefined type
+          // https://github.com/microsoft/TypeScript/blob/cb44488fcec4348a448434afbf2ebcbf2b423c61/src/compiler/checker.ts/#L2059
+          if (
+            caseTypes.has(checker.getUndefinedType()) &&
+            tsutils.isIntrinsicUndefinedType(intersectionPart)
+          ) {
+            continue;
+          }
+
```

Comment:
```suggestion
          // "missing", "optional" and "undefined" types are different runtime objects,
          // but all of them have TypeFlags.Undefined type flag
          if (
            Array.from(caseTypes).some(tsutils.isIntrinsicUndefinedType) &&
            tsutils.isIntrinsicUndefinedType(intersectionPart)
          ) {
            continue;
          }

```

[Bug] If we cannot rely `intersectionPart` being equal to `checker.getUndefinedType()`, we cannot rely on `caseTypes` to contain only `undefinedType` objects. `caseType` may also contain `missingType`:

```ts
function foo(x: string[], y: string | undefined) {
  const a = x[0];
  if (typeof a === 'string') {
    return
  }
  switch (y) {
    case 'hi':
      break;
    case a:
      break;
  }
}
```

[playground link](https://deploy-preview-10232--typescript-eslint.netlify.app/play/#ts=5.6.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUAPAXIgzlAJxjAHMBtAXQBpEBPHfI0xAH0XABMBTYYrjgJSIA3gChEiCAnyIAhogC8iDGQAMFANzjEMYIjRRaABy5w98hZcQByRsRLWhYiRIJcoIAmG0BfbbgB3GCgIAAt9WidtCQhZXC4bUJhrLGiXACM3WQBrLRdJOITZVPyJTK4cvMQ-PyA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJFwEipCtTqMWiAO5kcxABY0AHuth9EOMgDdkDIoiobkxANbDIqeAHs0PAL4g3QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false)

Reply:
Oh, nice catch, thanks!